### ISSUE TITLE
Column view when listing contents inside a folder

### ISSUE THREAD

**Issue by andreg** [Wramberg/TerminalView#51]
Hi,

when I <code>ls</code> the current folder, the column output doesn't look right.

![schermata 2017-08-07 alle 21 11 57](https://user-images.githubusercontent.com/39385/29041994-646bc9f6-7bb5-11e7-8e15-2c21c2ca9f50.png)

* OS: macOS 10.12.6
* ST3 version: Dev build 3142
* Bash version: 3.2.57(1)

Anyway, thanks for a great plugin!

**Comment by Wramberg** [comment]
I need some more of the information requested in the issue template.
What is the `cmd` argument you call TerminalView with ? Just bash -l ?
What is returned if you call `echo $TERM` inside the TerminalView ?

**Comment by andreg** [comment]
Calling `echo $TERM` returns `linux`. I'm not sure what you mean with the `cmd` argument question, but I haven't configured TerminalView in any specific way, it should be all factory defaults.

**Comment by rwols** [comment]
@andreg run the command
```
$ env | pbcopy
```
and press <kbd>command</kbd><kbd>V</kbd> here (remove any sensitive information)

**Comment by andreg** [comment]
There you go:

````
TERM_PROGRAM=Apple_Terminal
SHELL=/bin/bash
TERM=xterm-256color
TMPDIR=/var/folders/f9/rj3whr892l97gpv0ydrqszyr0000gn/T/
Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.rcLyiDcboK/Render
TERM_PROGRAM_VERSION=388.1.1
TERM_SESSION_ID=0891D9ED-FF8E-434C-A5EC-E28424395354
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.tL7NSkFDi9/Listeners
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0x4
PATH=~/vagrant-local/vv:/Applications/MAMP/bin/php/php7.1.1/bin:/Applications/MAMP/Library/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/git/bin:/usr/texbin
EDITOR=subl
LANG=it_IT.UTF-8
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
SHLVL=1
_=/usr/bin/env
````

**Comment by Wramberg** [comment]
DId you call `env | pbcopy` inside TerminalVIew ? If yes then it appears you have a .bash_profile, .basrc file or similar which changes the value of the TERM environment variable.

**Comment by andreg** [comment]
Sorry about that. Here's the output of the command when it's called from within TerminalView (I'm on a different machine right now, but I have the same problem):

````
TERM=linux
SHELL=/bin/bash
TMPDIR=/var/folders/w_/jvt95b454fd42cwkndf0nf6c0000gn/T/
Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.DZR4Ovs30J/Render
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.EPmSBtyEHn/Listeners
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0x4
PATH=~/vagrant-local/vv:/Applications/MAMP/bin/php/php7.1.1/bin:/Applications/MAMP/Library/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
EDITOR=subl
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
SHLVL=1
_=/usr/bin/env
````

**Comment by Wramberg** [comment]
THanks @andreg I'll look more into this. This is not intended behavior and definitely not something I've seen in any other tests. Do you have any thirdparty things installed like iTerm ? or anything else you can think of that would make your setup "out of the ordinary" ? Maybe some fancy ST3 theme ?

Might also be because you are running a very recent dev version of ST3. If you get the chance could you try the latest stable version ?

**Comment by andreg** [comment]
On my work machine, which is the one I've used to get you that last output, I'm using the latest stable version of Sublime.

I have tried the same command with both iTerm uninstalled and no custom ST3 theme, and the output is the same.

**Comment by Wramberg** [comment]
I have to admit this is a weird issue. I haven't tested with bash 3 but I doubt it's related to that. Could you list installed packages ? If you have alot I think it would be easiest to just call

```
ls $HOME/.config/sublime-text-text-3/Installed\ Packages
ls $HOME/.config/sublime-text-text-3/Packages
```

If you got time it might also be worth trying out the TerminalView 0.5.0 release candidate. You can install TerminalView manually like described in the README and just

```
git checkout 0.5.0
```

I'll dig more into your env variables later tonight - it's look OK but maybe I missed something.

**Comment by andreg** [comment]
Ok, these are the packages that I currently have installed:

````
0_package_control_loader.sublime-package
AdvancedNewFile.sublime-package
Alignment.sublime-package
All Autocomplete.sublime-package
ApacheConf.tmLanguage.sublime-package
Atomic Color Scheme.sublime-package
Boxy Theme.sublime-package
Codebug.sublime-package
DocBlockr.sublime-package
Emmet.sublime-package
Hugofy.sublime-package
Laravel Blade Highlighter.sublime-package
MarkdownEditing.sublime-package
Material Theme.sublime-package
Nettuts+ Fetch.sublime-package
Package Control.sublime-package
PackageResourceViewer.sublime-package
PyV8
Schemr.sublime-package
SideBarEnhancements.sublime-package
Sublime Tweet.sublime-package
SublimeLinter-php.sublime-package
SublimeOnSaveBuild.sublime-package
Syntax Highlighting for Sass.sublime-package
TOML.sublime-package
Theme - Soda.sublime-package
Theme - Spacegray.sublime-package
Toggle Words.sublime-package
ayu.sublime-package
````

I've tried with version 0.5.0, and I get the same output.
It is worth nothing that the output varies depending on how wide the window is:

![schermata 2017-08-08 alle 19 56 37](https://user-images.githubusercontent.com/39385/29086806-177e5676-7c74-11e7-8f39-03b18ed6f73b.png)

Thanks so much for spending time into this. It's not vital to have this fixed, but I think it's worth investigating.

**Comment by Wramberg** [comment]
I think its worth looking into since it may also cause other problems that are less obvious.

Couldn't replicate anything from your environment - seems good. Maybe @rwols would be so kind as to share his env for comparison (I dont have macOS) ?

Couldnt find any conflicting packages either.

Since it changes with view width it almost sounds like an old issue where wrapping wasnt turned off by the plugin causing some weird behavior. But in 0.4 and 0.5 it is specifically turned off by the plugin so this shouldnt be it. For safe measures you can try turning it on/off in the view. To do this you can add the following keybinding (or something similar) and hit it while the terminal view is active.

```
{ "keys": ["alt+q"], "command": "wrap_lines" },
```

It may be worthwhile trying to inspect your global ST3 settings to see if you have anything enabled that I didnt think could conflict with TerminalView so if you could post that it would be great.

Meanwhile I'll try downgrading my bash version but I think this is a long shot.

**Comment by andreg** [comment]
It could be OS-related then.

I've tried all of the above, and I've also tried emptying my settings and working with Sublime's default theme, and I still had the same output.

It is worth noting that different folders behave differently; depending more on the overall number of items rather than the length of their names, they tend to display wrong the more items you have in the folder.

**Comment by rwols** [comment]
OK. I have two laptops. One is a MacBook Pro with a retina screen and the other is an older MacBook Air from 2011 without a retina screen.

On the pro, I experience the same issue as @andreg if I use bash:
```
MBP-van-Raoul:CMakeBuilder raoulwols$ ls                                                                                                                                    
CMake.sublime-build            LICENSE.md                      __init__.py                                                                                                 m
essages                        tests                                                                                                                                        
CMakeBuilder.sublime-commands  Main.sublime-menu               commands                                                                                                    m
essages.json                   unittesting.json                                                                                                                             
CMakeBuilder.sublime-settings  README.md                       dependencies.json                                                                                           s
erver.py                                                                                                                                                                    
CONTRIBUTING.md                        Syntax                          generators                                                                                          s
upport    
```
However, fish does it right (this is why I haven't noticed this before):
```
~/L/A/S/P/CMakeBuilder (dev ⚡ =) ls                                                                                                                                         
CMake.sublime-build           LICENSE.md                    __init__.py                   messages                      tests                                               
CMakeBuilder.sublime-commands Main.sublime-menu             commands                      messages.json                 unittesting.json                                    
CMakeBuilder.sublime-settings README.md                     dependencies.json             server.py                                                                         
CONTRIBUTING.md               Syntax                        generators                    support       
```
**Moreover**, on the Air there are no problems with bash:
```
Air-van-Raoul:CMakeBuilder rwols$ ls                                                                                                       
CMake.sublime-build            Main.sublime-menu              dependencies.json              support/                                      
CMakeBuilder.sublime-commands  README.md                      generators/                    tests/                                        
CMakeBuilder.sublime-settings  Syntax/                        messages/                      unittesting.json                              
CONTRIBUTING.md                __init__.py                    messages.json                                                                
LICENSE.md                     commands/                      server.py   
```
Here are my env vars for the Pro:
```
Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.x2zxLHP9pX/Render
BROWSER=open -a /Applications/Safari.app
EDITOR=subl --wait
HOME=/Users/raoulwols
LC_CTYPE=en_US.UTF-8
LOGNAME=raoulwols
PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin
PWD=/Users/raoulwols/Library/Application Support/Sublime Text 3/Packages/CMakeBuilder
SHELL=/usr/local/bin/fish
SHLVL=2
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.W1f2oorIMV/Listeners
TERM=linux
TMPDIR=/var/folders/gr/22fzk32d28l4v5v7yvl4_bd80000gn/T/
USER=raoulwols
VISUAL=subl --wait
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0x5
```
And here they are for the Air:
```
TERM=linux
SHELL=/usr/local/bin/fish
CLICOLOR=1
TMPDIR=/var/folders/fc/w04z5z256yv_gm0tl5qx3mw00000gq/T/
Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.GJmVFVlrmH/Render
USER=rwols
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.9v6VlhZac6/Listeners
__CF_USER_TEXT_ENCODING=0x1F7:0x0:0x5
LSCOLORS=ExFxBxDxCxegedabagacad
PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Library/TeX/texbin
PWD=/Users/rwols/Library/Application Support/Sublime Text 3/Packages/CMakeBuilder
EDITOR=subl --wait
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
SHLVL=1
HOME=/Users/rwols
LOGNAME=rwols
VISUAL=subl --wait
LC_CTYPE=en_US.UTF-8
BROWSER=open -a /Applications/Safari.app
BASH_FUNC_makegif%%=() {  ffmpeg -i "$1" -s 1280x720 -r 30 -f gif - | gifsicle -O3 --delay=3 > "$2"
}
_=/usr/bin/env
OLDPWD=/Users/rwols

```

**Comment by Wramberg** [comment]
Oh no the problem is spreading ;-) I've borrowed a Macbook Pro from a buddy of mine - hopefully I'll be able to reproduce the issue.

**Comment by rwols** [comment]
With 4ddc273 the problem is gone here.

**Comment by Wramberg** [comment]
As it turns out vtxxx and linux has, by default, a tabstop for every 8 char. When starting pyte sets these tabstops

```
self.tabstops = set(range(7, self.columns, 8))
```

but when calling `tabs` in the terminal (command which reconfigures tabstops) they are essentially changed to

```
self.tabstops = set(range(8, self.columns, 8))
```

which begs the question if pyte has this wrong. With 4ddc2732d31b46e1486829d2f7bb6081ea829ef1 we now overwrite pyte default tabstops both on startup and on terminal reset. Furthermore we also adjust the tabstops when the terminal is resized larger - something pyte didnt do but assume needs to be done.

**Comment by Wramberg** [comment]
@rwols Thanks for the quick test :+1: 
@andreg Hopefully it is fixed for you as well - Ill set a little time aside to ensure this didnt break anything so expect it be released as 0.5.1 in the near future.

Just FYI @superbobry - maybe you have anything thoughts on this ?
It seems that bash 3.2.57(1) on macOS assumes default tabstops to be different from what pyte has as default.

**Comment by Wramberg** [comment]
Funny enough it seems that pymux has already learned the same lesson :-)

https://github.com/jonathanslenders/pymux/blob/9569dda0c74c5e8b1dd8563d7ecef2c184410db7/pymux/screen.py#L175
```
# (We choose to create tab stops until x=1000, because we keep the
# tab stops when the screen increases in size. The OS X 'ls' command
# relies on the stops to be there.)
self.tabstops = set(range(8, 1000, 8))
```

**Comment by andreg** [comment]
I can confirm all is good now, using the `devel` branch. Thank you so much!

**Comment by Wramberg** [comment]
Gonna close this for now. Hopefully Ill have time to release 0.5 soon. Seeing as pymux has done the same change as this one I'm gonna include the fix in 0.5 as well.
